# Starting point: GaussianNB classifier
GaussianNB()
	Accuracy: 0.82669	Precision: 0.33550	Recall: 0.12900	F1: 0.18635	F2: 0.14711
	Total predictions: 13000	True positives:  258	False positives:  511	False negatives: 1742	True negatives: 10489

In [3]:
#!/usr/bin/python

import sys
import pickle
import matplotlib.pyplot
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data
from sklearn.decomposition import PCA
from sklearn.svm import SVC, LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import train_test_split, cross_val_score
from sklearn.metrics import classification_report, precision_recall_curve
from sklearn.feature_selection import SelectKBest, chi2
from pprint import pprint
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".
features_list = ['poi','salary','bonus','total_payments'] # You will need to use more features

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)
print data_dict['METTS MARK']
### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        print e, data_dict[e]["salary"],  data_dict[e]['bonus']
        outliers.append(e)
print outliers
data_dict.pop('TOTAL',0)
#data_dict.pop('LAY KENNETH L',0)
#data_dict.pop('SKILLING JEFFREY K',0)
#       print e, data_dict[e]["salary"],  data_dict[e]['bonus']
#key_list = [k for k in data_dict.keys() if data_dict[k]["salary"] != 'NaN' and data_dict[k]["salary"] > 1000000 and data_dict[k]["bonus"] > 5000000]
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        print e, data_dict[e]["salary"],  data_dict[e]['bonus']
        outliers.append(e)
print outliers



### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)


### Task 4: Try a varity of classifiers
### Please name your classifier clf for easy export below.
### Note that if you want to do PCA or other multi-stage operations,
### you'll need to use Pipelines. For more info:
### http://scikit-learn.org/stable/modules/pipeline.html

# Provided to give you a starting point. Try a variety of classifiers.
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()

### Task 5: Tune your classifier to achieve better than .3 precision and recall
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info:  http://scikit-lea
### rn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSp
### lit.html

# Example starting point. Try investigating other evaluation techniques!
from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)

### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)
print " Features List ", features_list
test_classifier(clf, my_dataset, features_list)

{'salary': 365788, 'to_messages': 807, 'deferral_payments': 'NaN', 'total_payments': 1061827, 'exercised_stock_options': 'NaN', 'bonus': 600000, 'restricted_stock': 585062, 'shared_receipt_with_poi': 702, 'restricted_stock_deferred': 'NaN', 'total_stock_value': 585062, 'expenses': 94299, 'loan_advances': 'NaN', 'from_messages': 29, 'other': 1740, 'from_this_person_to_poi': 1, 'poi': False, 'director_fees': 'NaN', 'deferred_income': 'NaN', 'long_term_incentive': 'NaN', 'email_address': 'mark.metts@enron.com', 'from_poi_to_this_person': 38}
LAY KENNETH L 1072321 7000000
SKILLING JEFFREY K 1111258 5600000
TOTAL 26704229 97343619
['LAY KENNETH L', 'SKILLING JEFFREY K', 'TOTAL']
LAY KENNETH L 1072321 7000000
SKILLING JEFFREY K 1111258 5600000
['LAY KENNETH L', 'SKILLING JEFFREY K']
 Features List  ['poi', 'salary', 'bonus', 'total_payments']
GaussianNB()
	Accuracy: 0.82669	Precision: 0.33550	Recall: 0.12900	F1: 0.18635	F2: 0.14711
	Total predictions: 13000	True positives:  258	False positiv

In [4]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".
features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] # You will need to use more features

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        

data_dict.pop('TOTAL',0)
#data_dict.pop('LAY KENNETH L',0)
#data_dict.pop('SKILLING JEFFREY K',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.
for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})

# I replaced all 'NaN' values with '0' in order to clean up data and avoid any problem on the calcultions.

for ele in my_dataset:
    for e in features_list:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            my_dataset[ele][e] = 0 

### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)



### Task 4: Try a varity of classifiers
### Please name your classifier clf for easy export below.
### Note that if you want to do PCA or other multi-stage operations,
### you'll need to use Pipelines. For more info:
### http://scikit-learn.org/stable/modules/pipeline.html

# Provided to give you a starting point. Try a variety of classifiers.
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()

### Task 5: Tune your classifier to achieve better than .3 precision and recall
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info:  http://scikit-lea
### rn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSp
### lit.html

# Example starting point. Try investigating other evaluation techniques!
from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)

### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)
print "New features_list with  'ave_earnings':\n", features_list
test_classifier(clf, my_dataset, features_list)

New features_list with  'ave_earnings':
['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 'deferred_income', 'total_stock_value', 'exercised_stock_options', 'restricted_stock', 'restricted_stock_deferred', 'expenses', 'long_term_incentive', 'shared_receipt_with_poi', 'from_this_person_to_poi', 'from_poi_to_this_person', 'to_messages', 'from_messages']
GaussianNB()
	Accuracy: 0.78440	Precision: 0.28768	Recall: 0.41800	F1: 0.34081	F2: 0.38328
	Total predictions: 15000	True positives:  836	False positives: 2070	False negatives: 1164	True negatives: 10930



In [6]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".
features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] # You will need to use more features

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        

data_dict.pop('TOTAL',0)
#data_dict.pop('LAY KENNETH L',0)
#data_dict.pop('SKILLING JEFFREY K',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.
for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})

# I replaced all 'NaN' values with '0' in order to clean up data and avoid any problem on the calcultions.

for ele in my_dataset:
    for e in features_list:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            my_dataset[ele][e] = 0 

### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)



### Task 4: Try a varity of classifiers
### Please name your classifier clf for easy export below.
### Note that if you want to do PCA or other multi-stage operations,
### you'll need to use Pipelines. For more info:
### http://scikit-learn.org/stable/modules/pipeline.html

# Provided to give you a starting point. Try a variety of classifiers.
from sklearn.svm import SVC
clf = SVC(kernel='rbf', C=1000,gamma=.001)



### Task 5: Tune your classifier to achieve better than .3 precision and recall
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info:  http://scikit-lea
### rn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSp
### lit.html

# Example starting point. Try investigating other evaluation techniques!
from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.2, random_state=42)

    
from sklearn.metrics import classification_report
clf = clf.fit(features_train, labels_train)
accuracy =clf.score(features_test,labels_test)
pred = clf.predict(features_test)
target_names = ["Not POI", "POI"]
print "accuracy:",accuracy
print classification_report(y_true=labels_test, y_pred=pred, target_names=target_names)
        
### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)
print "New features_list with 'ave_earnings':\n", features_list
test_classifier(clf, my_dataset, features_list)

accuracy: 0.931034482759
             precision    recall  f1-score   support

    Not POI       0.93      1.00      0.96        27
        POI       0.00      0.00      0.00         2

avg / total       0.87      0.93      0.90        29

New features_list with 'ave_earnings':
['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 'deferred_income', 'total_stock_value', 'exercised_stock_options', 'restricted_stock', 'restricted_stock_deferred', 'expenses', 'long_term_incentive', 'shared_receipt_with_poi', 'from_this_person_to_poi', 'from_poi_to_this_person', 'to_messages', 'from_messages']
Got a divide by zero when trying out: SVC(C=1000, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.001, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
Precision or recall may be undefined due to a lack of true positive predicitons.


In [8]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".
features_list = ['poi','salary','bonus','total_payments'] # You will need to use more features

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        print e, data_dict[e]["salary"],  data_dict[e]['bonus']
        outliers.append(e)
        
print "outliers_before_removal :",outliers

data_dict.pop('TOTAL',0)
#data_dict.pop('LAY KENNETH L',0)
#data_dict.pop('SKILLING JEFFREY K',0)

outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        print e, data_dict[e]["salary"],  data_dict[e]['bonus']
        outliers.append(e)
print "outliers_after_removal :",outliers

### Task 3: Create new feature(s)


### Store to my_dataset for easy export below.
my_dataset = data_dict

### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)


### Task 4: Try a varity of classifiers
### Please name your classifier clf for easy export below.
### Note that if you want to do PCA or other multi-stage operations,
### you'll need to use Pipelines. For more info:
### http://scikit-learn.org/stable/modules/pipeline.html

# Provided to give you a starting point. Try a variety of classifiers.
from sklearn.svm import SVC
clf = SVC(kernel='rbf', C=1)


### Task 5: Tune your classifier to achieve better than .3 precision and recall
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info:  http://scikit-lea
### rn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSp
### lit.html

# Example starting point. Try investigating other evaluation techniques!
from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)

from sklearn.metrics import classification_report
clf = clf.fit(features_train, labels_train)
accuracy =clf.score(features_test,labels_test)
pred = clf.predict(features_test)
target_names = ["Not POI", "POI"]
print "accuracy:",accuracy
print classification_report(y_true=labels_test, y_pred=pred, target_names=target_names)
    
    
### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)
print " Features List ", features_list
test_classifier(clf, my_dataset, features_list)

LAY KENNETH L 1072321 7000000
SKILLING JEFFREY K 1111258 5600000
TOTAL 26704229 97343619
outliers_before_removal : ['LAY KENNETH L', 'SKILLING JEFFREY K', 'TOTAL']
LAY KENNETH L 1072321 7000000
SKILLING JEFFREY K 1111258 5600000
outliers_after_removal : ['LAY KENNETH L', 'SKILLING JEFFREY K']
accuracy: 0.789473684211
             precision    recall  f1-score   support

    Not POI       0.79      1.00      0.88        30
        POI       0.00      0.00      0.00         8

avg / total       0.62      0.79      0.70        38

 Features List  ['poi', 'salary', 'bonus', 'total_payments']
Got a divide by zero when trying out: SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
Precision or recall may be undefined due to a lack of true positive predicitons.


# Features List Data Cleaning
The script below is going to finalized features_list for future use in different identifier. The features_list contains selected features from original set of features in the data_dict, plus new feature, 'ave_earnings', and all 'NaN' values replaced with 0 on numerical features.

In [9]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".


features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        

data_dict.pop('TOTAL',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.
for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})

# I replaced all 'NaN' values with '0' in order to clean up data and avoid any problem on the calcultions.

for ele in my_dataset:
    for e in features_list:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            my_dataset[ele][e] = 0 

### Extract features and labels from dataset for local testing
from sklearn.cross_validation import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn import preprocessing
from sklearn.metrics import classification_report
from sklearn.naive_bayes import GaussianNB
from sklearn.feature_selection import SelectKBest
from sklearn.pipeline import Pipeline 


data = featureFormat(my_dataset, features_list, sort_keys = True)
labels, features = targetFeatureSplit(data)            
scaler = preprocessing.MinMaxScaler()
features = scaler.fit_transform(features)

print "Features List:", features_list



Features List: ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 'deferred_income', 'total_stock_value', 'exercised_stock_options', 'restricted_stock', 'restricted_stock_deferred', 'expenses', 'long_term_incentive', 'shared_receipt_with_poi', 'from_this_person_to_poi', 'from_poi_to_this_person', 'to_messages', 'from_messages']


# Feature Scaling and KBest data selection

The features_list will be scaled with MinMaxScaler and will go through KBest data selection in different identifiers. Below the selected features after scaling will be shown for K=1 thru 7 which are going to be used in our pipeline procedures for different identifers.

In [10]:



feature_train, feature_test, label_train, label_test = train_test_split(
features, labels, test_size=0.2,random_state=42)
kbest = SelectKBest(k=7)
kbest.fit_transform(feature_train, label_train)
features_new =[] 
for i in range(len(kbest.get_support())):
    if kbest.get_support().tolist()[i] is True:
        features_new.append(features_list[i+1])
features_list_new = ["poi"] + features_new
print "List of KBest selected features_list K=7 :\n",features_list_new

List of KBest selected features_list K=7 :
['poi', 'salary', 'bonus', 'deferred_income', 'total_stock_value', 'exercised_stock_options', 'long_term_incentive', 'shared_receipt_with_poi']


In [11]:



feature_train, feature_test, label_train, label_test = train_test_split(
features, labels, test_size=0.2,random_state=42)
kbest = SelectKBest(k=6)
kbest.fit_transform(feature_train, label_train)
features_new =[] 
for i in range(len(kbest.get_support())):
    if kbest.get_support().tolist()[i] is True:
        features_new.append(features_list[i+1])
features_list_new = ["poi"] + features_new
print "List of KBest selected features_list K=6 :\n",features_list_new

List of KBest selected features_list K=6 :
['poi', 'salary', 'bonus', 'total_stock_value', 'exercised_stock_options', 'long_term_incentive', 'shared_receipt_with_poi']


In [12]:


feature_train, feature_test, label_train, label_test = train_test_split(
features, labels, test_size=0.2,random_state=42)
kbest = SelectKBest(k=5)
kbest.fit_transform(feature_train, label_train)
features_new =[] 
for i in range(len(kbest.get_support())):
    if kbest.get_support().tolist()[i] is True:
        features_new.append(features_list[i+1])
features_list_new = ["poi"] + features_new
print "List of KBest selected features_list K=5 :\n",features_list_new

List of KBest selected features_list K=5 :
['poi', 'salary', 'bonus', 'total_stock_value', 'exercised_stock_options', 'long_term_incentive']


In [13]:


feature_train, feature_test, label_train, label_test = train_test_split(
features, labels, test_size=0.2,random_state=42)
kbest = SelectKBest(k=4)
kbest.fit_transform(feature_train, label_train)
features_new =[] 
for i in range(len(kbest.get_support())):
    if kbest.get_support().tolist()[i] is True:
        features_new.append(features_list[i+1])
features_list_new = ["poi"] + features_new
print "List of KBest selected features_list  K=4 :\n",features_list_new

List of KBest selected features_list  K=4 :
['poi', 'salary', 'bonus', 'total_stock_value', 'exercised_stock_options']


In [14]:


feature_train, feature_test, label_train, label_test = train_test_split(
features, labels, test_size=0.2,random_state=42)
kbest = SelectKBest(k=3)
kbest.fit_transform(feature_train, label_train)
features_new =[] 
for i in range(len(kbest.get_support())):
    if kbest.get_support().tolist()[i] is True:
        features_new.append(features_list[i+1])
features_list_new = ["poi"] + features_new
print "List of KBest selected features_list  K=3 :\n",features_list_new

List of KBest selected features_list  K=3 :
['poi', 'salary', 'bonus', 'total_stock_value']


In [15]:


feature_train, feature_test, label_train, label_test = train_test_split(
features, labels, test_size=0.2,random_state=42)
kbest = SelectKBest(k=2)
kbest.fit_transform(feature_train, label_train)
features_new =[] 
for i in range(len(kbest.get_support())):
    if kbest.get_support().tolist()[i] is True:
        features_new.append(features_list[i+1])
features_list_new = ["poi"] + features_new
print "List of KBest selected features_list  K=2 :\n",features_list_new

List of KBest selected features_list  K=2 :
['poi', 'salary', 'bonus']


In [16]:


feature_train, feature_test, label_train, label_test = train_test_split(
features, labels, test_size=0.2,random_state=42)
kbest = SelectKBest(k=1)
kbest.fit_transform(feature_train, label_train)
features_new =[] 
for i in range(len(kbest.get_support())):
    if kbest.get_support().tolist()[i] is True:
        features_new.append(features_list[i+1])
features_list_new = ["poi"] + features_new
print "List of KBest selected features_list  K=1 :\n",features_list_new

List of KBest selected features_list  K=1 :
['poi', 'bonus']


# Pipeline method with Scaler, SelectKBest, and GaussianNB ()
Comparing the precision result for identifiers configured Pipeline method with Scaler, SelectKBest, and GaussianNB () with K=4,5,and 6 values:

In [17]:
###Pipeline method with Scaler, SelectKBest (K=6), and GaussianNB () 

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=6)
gnb = GaussianNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("NaiveBayes", gnb)])


dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=6, score_func=<function f_classif at 0x085A82F0>)), ('NaiveBayes', GaussianNB())])
	Accuracy: 0.84620	Precision: 0.39921	Recall: 0.30400	F1: 0.34516	F2: 0.31923
	Total predictions: 15000	True positives:  608	False positives:  915	False negatives: 1392	True negatives: 12085



In [18]:
###Pipeline method with Scaler, SelectKBest (K=5), and GaussianNB () 


data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=5)
gnb = GaussianNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("NaiveBayes", gnb)])

dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=5, score_func=<function f_classif at 0x085A82F0>)), ('NaiveBayes', GaussianNB())])
	Accuracy: 0.85300	Precision: 0.43492	Recall: 0.34250	F1: 0.38322	F2: 0.35770
	Total predictions: 15000	True positives:  685	False positives:  890	False negatives: 1315	True negatives: 12110



In [19]:

###Pipeline method with Scaler, SelectKBest (K=5), and GaussianNB () 


data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=4)
gnb = GaussianNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("NaiveBayes", gnb)])

dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=4, score_func=<function f_classif at 0x085A82F0>)), ('NaiveBayes', GaussianNB())])
	Accuracy: 0.85033	Precision: 0.42152	Recall: 0.32900	F1: 0.36956	F2: 0.34411
	Total predictions: 15000	True positives:  658	False positives:  903	False negatives: 1342	True negatives: 12097



# Best result for Pipeline method with Scaler, SelectKBest, and GaussianNB () is for k=5 with Precision: 0.43492








# 

# 

# 

# 

# Pipeline method with Scaler, SelectKBest, and MultinomialNB()
Comparing the precision result for identifiers configured Pipeline method with Scaler, SelectKBest, and MultinomialNB() with K=2 thru 6 values:


# Best result for Pipeline method with Scaler, SelectKBest, and MultinomialNB():

Precision: 1.00000 is obtainable for K values from 3 to 5 but they seem too good to be true.

List of KBest selected features_list  K=6 ['poi', 'salary', 'bonus', 'deferred_income', 'total_stock_value', 'exercised_stock_options', 'expenses']
 Precision: 0.90741







In [20]:




###Pipeline method with Scaler, SelectKBest (K=2), and MultinomialNB () 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.3,random_state=42)
skb = SelectKBest(k=3)
mnb = MultinomialNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("MultinomialNB", mnb)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=3, score_func=<function f_classif at 0x085A82F0>)), ('MultinomialNB', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))])
	Accuracy: 0.86793	Precision: 1.00000	Recall: 0.00950	F1: 0.01882	F2: 0.01185
	Total predictions: 15000	True positives:   19	False positives:    0	False negatives: 1981	True negatives: 13000



In [21]:




###Pipeline method with Scaler, SelectKBest (K=5), and MultinomialNB () 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=5)
mnb = MultinomialNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("MultinomialNB", mnb)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=5, score_func=<function f_classif at 0x085A82F0>)), ('MultinomialNB', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))])
	Accuracy: 0.87247	Precision: 1.00000	Recall: 0.04350	F1: 0.08337	F2: 0.05379
	Total predictions: 15000	True positives:   87	False positives:    0	False negatives: 1913	True negatives: 13000



In [22]:




###Pipeline method with Scaler, SelectKBest (K=6), and MultinomialNB () 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=6)
mnb = MultinomialNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("MultinomialNB", mnb)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=6, score_func=<function f_classif at 0x085A82F0>)), ('MultinomialNB', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))])
	Accuracy: 0.87253	Precision: 0.90741	Recall: 0.04900	F1: 0.09298	F2: 0.06043
	Total predictions: 15000	True positives:   98	False positives:   10	False negatives: 1902	True negatives: 12990



In [23]:




###Pipeline method with Scaler, SelectKBest (K=7), and MultinomialNB () 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=7)
mnb = MultinomialNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("MultinomialNB", mnb)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=7, score_func=<function f_classif at 0x085A82F0>)), ('MultinomialNB', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))])
	Accuracy: 0.87173	Precision: 0.79688	Recall: 0.05100	F1: 0.09586	F2: 0.06275
	Total predictions: 15000	True positives:  102	False positives:   26	False negatives: 1898	True negatives: 12974



# Pipeline method with Scaler, SelectKBest, and SVC()
Comparing the precision result for identifiers configured Pipeline method with Scaler, SelectKBest, and SVC() with K=1 thru 4 values:

In [24]:
from sklearn.svm import SVC



###Pipeline method with Scaler, SelectKBest (K=4), and svm() 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=1)
svm = SVC(kernel='rbf', C=1)
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=1, score_func=<function f_classif at 0x085A82F0>)), ('SVC', SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.86680	Precision: 0.50694	Recall: 0.03650	F1: 0.06810	F2: 0.04482
	Total predictions: 15000	True positives:   73	False positives:   71	False negatives: 1927	True negatives: 12929



In [25]:
from sklearn.svm import SVC



###Pipeline method with Scaler, SelectKBest (K=4), and svm() 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=2)
svm = SVC(kernel='rbf', C=1)
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=2, score_func=<function f_classif at 0x085A82F0>)), ('SVC', SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.87080	Precision: 0.84444	Recall: 0.03800	F1: 0.07273	F2: 0.04697
	Total predictions: 15000	True positives:   76	False positives:   14	False negatives: 1924	True negatives: 12986



In [27]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data
from sklearn.decomposition import PCA
from sklearn.svm import SVC, LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import train_test_split
from sklearn.metrics import classification_report, precision_recall_curve
from sklearn.feature_selection import SelectKBest, chi2
from pprint import pprint
sys.path.append("../tools/")


### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".


features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
# bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        
print "Outliers Before Removal of TOTAL :",outliers

data_dict.pop('TOTAL',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.

my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.

for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})

print 'ave_earnings is the average value of:', features_list[1:5]

      
       
###Extract features and labels from dataset for local testing
# I removed entries with all 'NaN' values or all '0' in order to clean up data and avoid any problem on calcultions.
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)
print "\n Features List:\n"
pprint (features_list)



### Task 4: Try a varity of classifiers

# I tried several different classifiers and their output results of the tester.py script
# in the ML_project_varity_of_classifiers.ipny file
# Provided to give you a starting point. Try a variety of classifiers.
#from sklearn.naive_bayes import GaussianNB
#from sklearn import cross_validation



### Task 5: Tune your classifier to achieve better than .3 precision and recall 
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info: 
### http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSplit.html

# Below is the clasifier with best precision result above .3




data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=2)
svm = SVC(kernel='rbf', C=1)
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Outliers Before Removal of TOTAL : ['LAY KENNETH L', 'SKILLING JEFFREY K', 'TOTAL']
ave_earnings is the average value of: ['salary', 'bonus', 'deferral_payments', 'total_payments']

 Features List:

['poi',
 'salary',
 'bonus',
 'deferral_payments',
 'total_payments',
 'ave_earnings',
 'deferred_income',
 'total_stock_value',
 'exercised_stock_options',
 'restricted_stock',
 'restricted_stock_deferred',
 'expenses',
 'long_term_incentive',
 'shared_receipt_with_poi',
 'from_this_person_to_poi',
 'from_poi_to_this_person',
 'to_messages',
 'from_messages']
Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=2, score_func=<function f_classif at 0x085A82F0>)), ('SVC', SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.87080	Precision: 0.84444	Recall: 0.03800	F

In [28]:
from sklearn.svm import SVC



###Pipeline method with Scaler, SelectKBest (K=4), and svm() 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=3)
svm = SVC(kernel='rbf', C=1)
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=3, score_func=<function f_classif at 0x085A82F0>)), ('SVC', SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.86773	Precision: 0.78571	Recall: 0.01100	F1: 0.02170	F2: 0.01370
	Total predictions: 15000	True positives:   22	False positives:    6	False negatives: 1978	True negatives: 12994



In [29]:
from sklearn.svm import SVC



###Pipeline method with Scaler, SelectKBest (K=4), and svm() 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=4)
svm = SVC(kernel='rbf', C=1)
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=4, score_func=<function f_classif at 0x085A82F0>)), ('SVC', SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.86693	Precision: 0.75000	Recall: 0.00300	F1: 0.00598	F2: 0.00375
	Total predictions: 15000	True positives:    6	False positives:    2	False negatives: 1994	True negatives: 12998



In [30]:
from sklearn.svm import SVC



###Pipeline method with Scaler, SelectKBest (K=4), and svm() 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=2)
svm = SVC(kernel='rbf',C=1.000, max_iter=1000,degree=3,gamma=1)
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=2, score_func=<function f_classif at 0x085A82F0>)), ('SVC', SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=1, kernel='rbf',
  max_iter=1000, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.86673	Precision: 0.50276	Recall: 0.04550	F1: 0.08345	F2: 0.05562
	Total predictions: 15000	True positives:   91	False positives:   90	False negatives: 1909	True negatives: 12910



In [32]:
from sklearn.svm import SVC



###Pipeline method with Scaler, SelectKBest (K=4), and svm() 
from sklearn.naive_bayes import MultinomialNB

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=2)
svm = SVC(C=1000, max_iter=10000,degree=1, kernel='linear',gamma="auto")
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])



dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=2, score_func=<function f_classif at 0x085A82F0>)), ('SVC', SVC(C=1000, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=1, gamma='auto', kernel='linear',
  max_iter=10000, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.87133	Precision: 0.63060	Recall: 0.08450	F1: 0.14903	F2: 0.10220
	Total predictions: 15000	True positives:  169	False positives:   99	False negatives: 1831	True negatives: 12901



# Best result for Pipeline method with Scaler, SelectKBest, and SVC () is :

List of KBest selected features_list  K=2 :['poi', 'total_stock_value', 'exercised_stock_options']
Precision: 0.84444  








# 

# PCA
I will check PCA with different parameters to find the best features among 18 seleted features in the dataset.

In [33]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".


features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        

data_dict.pop('TOTAL',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.
for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})





### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)

def doPCA():
    from sklearn.decomposition import PCA
    pca = PCA(n_components=None, copy=True, whiten=False)
    pca.fit(data)
    return pca

pca = doPCA()
print pca.explained_variance_ratio_




[  7.93497604e-01   1.62548991e-01   1.91840911e-02   1.00914934e-02
   7.19056441e-03   4.22065416e-03   2.02841480e-03   1.02980131e-03
   1.12341912e-04   6.15343671e-05   2.35941988e-05   1.08774198e-05
   2.83270452e-08   8.91942962e-09   7.94354765e-10   1.77275871e-11
   1.30617092e-11   4.66936034e-16]


# Best parameters set found on SVC development set:
()
{'kernel': 'rbf', 'C': 100, 'gamma': 1}

In [35]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data
from sklearn.decomposition import PCA
from sklearn.svm import SVC, LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import train_test_split
from sklearn.metrics import classification_report, precision_recall_curve
from sklearn.feature_selection import SelectKBest, chi2
from pprint import pprint
sys.path.append("../tools/")


### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".


features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
# bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        
print outliers

data_dict.pop('TOTAL',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.

my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.

for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})

print 'ave_earnings is the average value of', features_list[1:5]

      
       
###Extract features and labels from dataset for local testing
# I removed entries with all 'NaN' values or all '0' in order to clean up data and avoid any problem on calcultions.
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)
print "\n Features List:\n"
pprint (features_list)



### Task 4: Try a varity of classifiers

# I tried several different classifiers and their output results of the tester.py script
# in the ML_project_varity_of_classifiers.ipny file
# Provided to give you a starting point. Try a variety of classifiers.
#from sklearn.naive_bayes import GaussianNB
#from sklearn import cross_validation



### Task 5: Tune your classifier to achieve better than .3 precision and recall 
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info: 
### http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSplit.html

# Below is the clasifier with best precision result above .3



feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.3,random_state=42)

#from __future__ import print_function


from sklearn.cross_validation import train_test_split
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import classification_report
from sklearn.svm import SVC

print(__doc__)

# Set the parameters by cross-validation
tuned_parameters = [{'kernel': ['rbf'], 'gamma': [1, 1e-1, 1e-3, 1e-4],'C': [100,1000, 10000]}]

scores = ['precision', 'recall']

for score in scores:
    print("# Tuning hyper-parameters for %s" % score)
    print()
    clf = GridSearchCV(SVC(), tuned_parameters, cv=5,scoring='%s_weighted' % score)
    clf.fit(feature_train, label_train)
    print("Best parameters set found on development set:")
    print()
    print(clf.best_params_)
    print()
    print("Grid scores on development set:")
    print()
    for params, mean_score, scores in clf.grid_scores_:
        print("%0.3f (+/-%0.03f) for %r"
              % (mean_score, scores.std() * 2, params))
    print()
    print("Detailed classification report:")
    print()
    print("The model is trained on the full development set.")
    print("The scores are computed on the full evaluation set.")
    print()
    label_true, label_pred = label_test, clf.predict(feature_test)
    print(classification_report(label_true, label_pred))
    print()



['LAY KENNETH L', 'SKILLING JEFFREY K', 'TOTAL']
ave_earnings is the average value of ['salary', 'bonus', 'deferral_payments', 'total_payments']

 Features List:

['poi',
 'salary',
 'bonus',
 'deferral_payments',
 'total_payments',
 'ave_earnings',
 'deferred_income',
 'total_stock_value',
 'exercised_stock_options',
 'restricted_stock',
 'restricted_stock_deferred',
 'expenses',
 'long_term_incentive',
 'shared_receipt_with_poi',
 'from_this_person_to_poi',
 'from_poi_to_this_person',
 'to_messages',
 'from_messages']
 a basic script for importing student's POI identifier,
    and checking the results that they get from it 
 
    requires that the algorithm, dataset, and features list
    be written to my_classifier.pkl, my_dataset.pkl, and
    my_feature_list.pkl, respectively

    that process should happen at the end of poi_id.py

# Tuning hyper-parameters for precision
()
Best parameters set found on development set:
()
{'kernel': 'rbf', 'C': 100, 'gamma': 1}
()
Grid scores on de

The PCA with default parametr values, returns the percentage of variance explained by all of the components. In next examination I try n_components=4 for the first four componenets.

In [36]:
### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)

def doPCA():
    from sklearn.decomposition import PCA
    pca = PCA(n_components=4, copy=True, whiten=False)
    pca.fit(data)
    return pca

pca = doPCA()
print pca.explained_variance_ratio_




[ 0.7934976   0.16254899  0.01918409  0.01009149]


In [37]:
### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)

def doPCA():
    from sklearn.decomposition import PCA
    pca = PCA(n_components=4, copy=True, whiten=True)
    pca.fit(data)
    return pca

pca = doPCA()
print pca.explained_variance_ratio_




[ 0.7934976   0.16254899  0.01918409  0.01009149]


Changing the the default values for (copy=True, whiten=True) did not change the result and the explained_variance_ratio_ was the same. In the next steps I will try different combinations of PCA and clasifiers for n_components=2 as they have highest explained_variance_ratio.

# Pipeline with PCA and GaussianNB()



In [38]:


from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.decomposition import PCA

estimators = [('reduce_dim', PCA(n_components=2)), ('gaussiannb', GaussianNB())]
clf = Pipeline(estimators)
 
    
### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)  

Pipeline(steps=[('reduce_dim', PCA(copy=True, n_components=2, whiten=False)), ('gaussiannb', GaussianNB())])
	Accuracy: 0.87073	Precision: 0.52837	Recall: 0.28400	F1: 0.36943	F2: 0.31295
	Total predictions: 15000	True positives:  568	False positives:  507	False negatives: 1432	True negatives: 12493



In [39]:


from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.decomposition import PCA

estimators = [('reduce_dim', PCA(n_components=1)), ('gaussiannb', GaussianNB())]
clf = Pipeline(estimators)
 
    
### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)  

Pipeline(steps=[('reduce_dim', PCA(copy=True, n_components=1, whiten=False)), ('gaussiannb', GaussianNB())])
	Accuracy: 0.87860	Precision: 0.61854	Recall: 0.23350	F1: 0.33902	F2: 0.26670
	Total predictions: 15000	True positives:  467	False positives:  288	False negatives: 1533	True negatives: 12712



In [40]:


from sklearn.naive_bayes import GaussianNB
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA

estimators = [('reduce_dim', PCA(n_components=4)), ('gaussiannb', GaussianNB())]
clf = Pipeline(estimators)
 
    
### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)  

Pipeline(steps=[('reduce_dim', PCA(copy=True, n_components=4, whiten=False)), ('gaussiannb', GaussianNB())])
	Accuracy: 0.86513	Precision: 0.48891	Recall: 0.25350	F1: 0.33388	F2: 0.28051
	Total predictions: 15000	True positives:  507	False positives:  530	False negatives: 1493	True negatives: 12470



The Best result for pipeline with PCA (n_components=1) and GaussianNB():
Precision: 0.61854

# Pipeline with PCA and SVM
I will use pipeline to chain PCA and SVM for the identifier and then will use grid search to see the effect of different parameter setup for each one of them onthe output.


In [41]:
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.decomposition import PCA
from sklearn.grid_search import GridSearchCV

estimators = [('reduce_dim', PCA()), ('svm', SVC(C=1000,gamma=.001))]
clf = Pipeline(estimators)
clf
#SVC(C=1.0, kernel='rbf', degree=3, gamma='auto', coef0=0.0, shrinking=True, probability=False, 
#tol=0.001, cache_size=200, class_weight=None, verbose=False, max_iter=-1, decision_function_shape=None, random_state=None)

  

dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)  

Got a divide by zero when trying out: Pipeline(steps=[('reduce_dim', PCA(copy=True, n_components=None, whiten=False)), ('svm', SVC(C=1000, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.001, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
Precision or recall may be undefined due to a lack of true positive predicitons.


In [ ]:
import pickle
import sys
import matplotlib.pyplot
sys.path.append("../tools/")
from feature_format import featureFormat, targetFeatureSplit
from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### read in data dictionary, convert to numpy array
data_dict = pickle.load( open("../final_project/final_project_dataset.pkl", "r") )

features_list = ['salary', 'bonus', 'deferral_payments', 'total_payments', 
                 'deferred_income','total_stock_value']
#data = featureFormat(data_dict, features)
#print data

outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        

data_dict.pop('TOTAL',0)

my_dataset = data_dict

### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True, remove_NaN=True,
                     remove_all_zeroes=True, remove_any_zeroes=False)
labels, features = targetFeatureSplit(data)


### your code below
for point in data:
    salary = point[0]
    bonus = point[1]
    matplotlib.pyplot.scatter( salary, bonus )

matplotlib.pyplot.xlabel("salary")
matplotlib.pyplot.ylabel("bonus")
matplotlib.pyplot.show()


from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.decomposition import PCA
estimators = [('reduce_dim', PCA(n_components=2)), ('svm', SVC(kernel='rbf', C=1))]
clf = Pipeline(estimators)
print clf 



import matplotlib.pyplot as plt
def doPCA():
    from sklearn.decomposition import PCA
    pca = PCA(n_components=2)
    pca.fit(data)
    return pca

pca = doPCA()
print pca.explained_variance_ratio_
first_pc = pca.components_[0]
second_pc = pca.components_[1]

transformed_data = pca.transform(data)
for ii, jj in zip(transformed_data, data):
    plt.scatter( first_pc[0]*ii[0], first_pc[1]*ii[0], color="r")
    plt.scatter( second_pc[0]*ii[1], second_pc[1]*ii[1], color="c")

    
plt.show()

#from sklearn import cross_validation
#X_train, X_test, y_train, y_test = cross_validation.train_test_split(features, labels, test_size=0.4, random_state=42)
#X_train[0], X_test[0], y_train[0], y_test[0]





# Example starting point. Try investigating other evaluation techniques!
from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)
    


### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)

#test_classifier(clf, my_dataset, features_list)

In [14]:

import matplotlib.pyplot as plt
def doPCA():
    from sklearn.decomposition import PCA
    pca = PCA(n_components=4)
    pca.fit(data)
    return pca

pca = doPCA()
print pca.explained_variance_ratio_
first_pc = pca.components_[0]
second_pc = pca.components_[1]


#from sklearn import cross_validation
#X_train, X_test, y_train, y_test = cross_validation.train_test_split(features, labels, test_size=0.4, random_state=42)
#X_train[0], X_test[0], y_train[0], y_test[0]





# Example starting point. Try investigating other evaluation techniques!
from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)
    


### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)

#test_classifier(clf, my_dataset, features_list)

[ 0.84148041  0.14221172  0.00931718  0.00536635]


In [11]:


from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.decomposition import PCA
estimators = [('reduce_dim', PCA(n_components=2)), ('svm', SVC(kernel='rbf', C=1))]
clf = Pipeline(estimators)
print clf 




from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.2, random_state=20)
    


### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)

test_classifier(clf, my_dataset, features_list)

Pipeline(steps=[('reduce_dim', PCA(copy=True, n_components=2, whiten=False)), ('svm', SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
Got a divide by zero when trying out: Pipeline(steps=[('reduce_dim', PCA(copy=True, n_components=2, whiten=False)), ('svm', SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
Precision or recall may be undefined due to a lack of true positive predicitons.


In [95]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".


features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        


data_dict.pop('TOTAL',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.
for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
#    print earnings, np.mean(earnings)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})

# I replaced all 'NaN' values with '0' in order to clean up data and avoid any problem on the calcultions.

for ele in my_dataset:
    for e in features_list:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            my_dataset[ele][e] = 0 

### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True)
labels, features = targetFeatureSplit(data)



### Task 4: Try a varity of classifiers
from sklearn.cross_validation import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report
from sklearn.naive_bayes import GaussianNB
from sklearn.feature_selection import SelectKBest
from sklearn.pipeline import Pipeline 


data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=5)
gnb = GaussianNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("NaiveBayes", gnb)])

# Provided to give you a starting point. Try a variety of classifiers.




### Task 5: Tune your classifier to achieve better than .3 precision and recall 
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info: 
### http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSplit.html

# Example starting point. Try investigating other evaluation techniques!

test_classifier(clf, my_dataset, features_list) 

### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but mgake sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=5, score_func=<function f_classif at 0x0562F070>)), ('NaiveBayes', GaussianNB())])
	Accuracy: 0.85300	Precision: 0.43492	Recall: 0.34250	F1: 0.38322	F2: 0.35770
	Total predictions: 15000	True positives:  685	False positives:  890	False negatives: 1315	True negatives: 12110



In [96]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".


features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        

data_dict.pop('TOTAL',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.
for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})




### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True)
labels, features = targetFeatureSplit(data)





### Task 4: Try a varity of classifiers
from sklearn.cross_validation import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_selection import SelectKBest
from sklearn.pipeline import Pipeline 


data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=6)
mnb = MultinomialNB()
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("MultinomialNB", mnb)])



### Task 5: Tune your classifier to achieve better than .3 precision and recall 
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info: 
### http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSplit.html

# Example starting point. Try investigating other evaluation techniques!


### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)
test_classifier(clf, my_dataset, features_list) 

Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=6, score_func=<function f_classif at 0x0562F070>)), ('MultinomialNB', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))])
	Accuracy: 0.87253	Precision: 0.90741	Recall: 0.04900	F1: 0.09298	F2: 0.06043
	Total predictions: 15000	True positives:   98	False positives:   10	False negatives: 1902	True negatives: 12990



In [100]:
#!/usr/bin/python

import sys
import pickle
import numpy as np
sys.path.append("../tools/")

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

### Task 1: Select what features you'll use.
### features_list is a list of strings, each of which is a feature name.
### The first feature must be "poi".


features_list = ['poi', 'salary', 'bonus', 'deferral_payments', 'total_payments', 'ave_earnings', 
                 'deferred_income','total_stock_value', 'exercised_stock_options', 
                'restricted_stock', 'restricted_stock_deferred', 'expenses',  
                 'long_term_incentive', 'shared_receipt_with_poi', 
                 'from_this_person_to_poi','from_poi_to_this_person',
                'to_messages','from_messages'] 

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

### Task 2: Remove outliers
# List of all keys of the data_dict for  salary value > 1 million and 
#bonus > 5 million dollars
outliers = []
for e in data_dict.keys():
    if data_dict[e]["salary"] != 'NaN' and data_dict[e]['salary'] > 1000000 and data_dict[e]['bonus'] > 5000000:
        outliers.append(e)
        

data_dict.pop('TOTAL',0)


### Task 3: Create new feature(s)
### Store to my_dataset for easy export below.
my_dataset = data_dict

#### I will add a new feature that shows average value of total earning called ave_earnings 
#### by calculating the mean value of 'salary', 'bonus', 'deferral_payments', and 'total_payments' for each person.
for ele in my_dataset:
    earnings = []
    for e in features_list[1:5]:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            earn = 0
            earnings.append(earn)
        earnings.append(earn)
#    print earnings, np.mean(earnings)
    ave_earnings = np.mean(earnings)
    my_dataset[ele].update({'ave_earnings': ave_earnings})

# I replaced all 'NaN' values with '0' in order to clean up data and avoid any problem on the calcultions.

for ele in my_dataset:
    for e in features_list:
        earn = my_dataset[ele][e]
        if earn =='NaN':
            my_dataset[ele][e] = 0 



### Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, features_list, sort_keys = True)
labels, features = targetFeatureSplit(data)




### Task 4: Try a varity of classifiers
from sklearn.cross_validation import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_selection import SelectKBest
from sklearn.pipeline import Pipeline 
from sklearn.svm import SVC
from sklearn.decomposition import PCA

data = featureFormat(my_dataset, features_list)
labels, features = targetFeatureSplit(data)
feature_train, feature_test, label_train, label_test = train_test_split( 
features, labels, test_size=0.2,random_state=42)
skb = SelectKBest(k=6)
svm = SVC(kernel='rbf', C=1000)
scaler = MinMaxScaler()
clf = Pipeline(steps=[('scaling',scaler),("SKB", skb), ("SVC", svm)])


# Provided to give you a starting point. Try a variety of classifiers.



#from sklearn import cross_validation



### Task 5: Tune your classifier to achieve better than .3 precision and recall 
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info: 
### http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSplit.html

# Example starting point. Try investigating other evaluation techniques!
test_classifier(clf, my_dataset, features_list) 

### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, features_list)


Pipeline(steps=[('scaling', MinMaxScaler(copy=True, feature_range=(0, 1))), ('SKB', SelectKBest(k=6, score_func=<function f_classif at 0x0562F070>)), ('SVC', SVC(C=1000, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False))])
	Accuracy: 0.85493	Precision: 0.34058	Recall: 0.09400	F1: 0.14734	F2: 0.10992
	Total predictions: 15000	True positives:  188	False positives:  364	False negatives: 1812	True negatives: 12636



In [2]:
#!/usr/bin/pickle

""" a basic script for importing student's POI identifier,
    and checking the results that they get from it 
 
    requires that the algorithm, dataset, and features list
    be written to my_classifier.pkl, my_dataset.pkl, and
    my_feature_list.pkl, respectively

    that process should happen at the end of poi_id.py
"""

import pickle
import sys
from sklearn.cross_validation import StratifiedShuffleSplit
sys.path.append("../tools/")
from feature_format import featureFormat, targetFeatureSplit

PERF_FORMAT_STRING = "\
\tAccuracy: {:>0.{display_precision}f}\tPrecision: {:>0.{display_precision}f}\t\
Recall: {:>0.{display_precision}f}\tF1: {:>0.{display_precision}f}\tF2: {:>0.{display_precision}f}"
RESULTS_FORMAT_STRING = "\tTotal predictions: {:4d}\tTrue positives: {:4d}\tFalse positives: {:4d}\
\tFalse negatives: {:4d}\tTrue negatives: {:4d}"

def test_classifier(clf, dataset, feature_list, folds = 1000):
    data = featureFormat(dataset, feature_list, sort_keys = True)
    labels, features = targetFeatureSplit(data)
    cv = StratifiedShuffleSplit(labels, folds, random_state = 42)
    true_negatives = 0
    false_negatives = 0
    true_positives = 0
    false_positives = 0
    for train_idx, test_idx in cv: 
        features_train = []
        features_test  = []
        labels_train   = []
        labels_test    = []
        for ii in train_idx:
            features_train.append( features[ii] )
            labels_train.append( labels[ii] )
        for jj in test_idx:
            features_test.append( features[jj] )
            labels_test.append( labels[jj] )
        
        ### fit the classifier using training set, and test on test set
        clf.fit(features_train, labels_train)
        predictions = clf.predict(features_test)
        for prediction, truth in zip(predictions, labels_test):
            if prediction == 0 and truth == 0:
                true_negatives += 1
            elif prediction == 0 and truth == 1:
                false_negatives += 1
            elif prediction == 1 and truth == 0:
                false_positives += 1
            elif prediction == 1 and truth == 1:
                true_positives += 1
            else:
                print "Warning: Found a predicted label not == 0 or 1."
                print "All predictions should take value 0 or 1."
                print "Evaluating performance for processed predictions:"
                break
    try:
        total_predictions = true_negatives + false_negatives + false_positives + true_positives
        accuracy = 1.0*(true_positives + true_negatives)/total_predictions
        precision = 1.0*true_positives/(true_positives+false_positives)
        recall = 1.0*true_positives/(true_positives+false_negatives)
        f1 = 2.0 * true_positives/(2*true_positives + false_positives+false_negatives)
        f2 = (1+2.0*2.0) * precision*recall/(4*precision + recall)
        print clf
        print PERF_FORMAT_STRING.format(accuracy, precision, recall, f1, f2, display_precision = 5)
        print RESULTS_FORMAT_STRING.format(total_predictions, true_positives, false_positives, false_negatives, true_negatives)
        print ""
    except:
        print "Got a divide by zero when trying out:", clf
        print "Precision or recall may be undefined due to a lack of true positive predicitons."

CLF_PICKLE_FILENAME = "my_classifier.pkl"
DATASET_PICKLE_FILENAME = "my_dataset.pkl"
FEATURE_LIST_FILENAME = "my_feature_list.pkl"

def dump_classifier_and_data(clf, dataset, feature_list):
    with open(CLF_PICKLE_FILENAME, "w") as clf_outfile:
        pickle.dump(clf, clf_outfile)
    with open(DATASET_PICKLE_FILENAME, "w") as dataset_outfile:
        pickle.dump(dataset, dataset_outfile)
    with open(FEATURE_LIST_FILENAME, "w") as featurelist_outfile:
        pickle.dump(feature_list, featurelist_outfile)

def load_classifier_and_data():
    with open(CLF_PICKLE_FILENAME, "r") as clf_infile:
        clf = pickle.load(clf_infile)
    with open(DATASET_PICKLE_FILENAME, "r") as dataset_infile:
        dataset = pickle.load(dataset_infile)
    with open(FEATURE_LIST_FILENAME, "r") as featurelist_infile:
        feature_list = pickle.load(featurelist_infile)
    return clf, dataset, feature_list

def main():
    ### load up student's classifier, dataset, and feature_list
    clf, dataset, feature_list = load_classifier_and_data()
    ### Run testing script
    test_classifier(clf, dataset, feature_list)

if __name__ == '__main__':
    main()


GaussianNB()
	Accuracy: 0.82669	Precision: 0.33550	Recall: 0.12900	F1: 0.18635	F2: 0.14711
	Total predictions: 13000	True positives:  258	False positives:  511	False negatives: 1742	True negatives: 10489

